blockchain-albatross/src/blockchain.rs

```diff
@@ -1239,72 +1220,56 @@ impl Blockchain {
             &read_txn
         };
 
+        // Gets slots collection from either the cached ones, or from the macro block.
         let slots_owned;
-        let slots;
-        if policy::epoch_at(state.block_number()) == policy::epoch_at(block_number) {
-            slots = state.current_slots.as_ref().expect("Missing current epoch's slots");
-        } else if policy::epoch_at(state.block_number()) == policy::epoch_at(block_number) + 1 {
-            slots = state.last_slots.as_ref().expect("Missing previous epoch's slots");
-        } else {
+        let slots = if policy::epoch_at(state.block_number()) == policy::epoch_at(block_number) {
+            state.current_slots.as_ref().expect("Missing current epoch's slots")
+        }
+        else if policy::epoch_at(state.block_number()) == policy::epoch_at(block_number) + 1 {
+            state.last_slots.as_ref()
+                .unwrap_or_else(|| panic!("Missing previous epoch's slots for block {}.{}", block_number, view_number))
+        }
+        else {
             let macro_block = self.chain_store
                 .get_block_at(policy::macro_block_before(block_number), true, Some(&txn))
-                .or_else(|| {
-                    warn!("Failed to determine slots - preceding macro block not found: block_number={}, view_number={}, state.block_number()={}", block_number, view_number, state.block_number());
-                    None
-                })?
+                .unwrap_or_else(|| panic!("Failed to determine slots - preceding macro block not found: block_number={}, view_number={}, state.block_number()={}", block_number, view_number, state.block_number()))
                 .unwrap_macro();
 
             // Get slots of epoch
             slots_owned = macro_block.try_into().unwrap();
-            slots = &slots_owned;
-        }
-
-        let idx_slot = state.reward_registry.slot_owner(block_number, view_number, slots, Some(&txn));
-        drop(state);
-
-        if idx_slot.is_some() {
-            return idx_slot;
-        }
-
-        // Try to get block producer from chain store
-        // if outside of range tracked by registry
-
-        let epoch_number = policy::epoch_at(block_number);
-        if block_number >= policy::first_block_of_registry(epoch_number) {
-            return None;
-        }
+            &slots_owned
+        };
 
-        self.chain_store.get_chain_info_at(block_number, false, Some(&txn))
-            .filter(|info| view_number == info.head.view_number())
-            .and_then(|info| info.slot)
+        state.reward_registry.get_slot_at(block_number, view_number, slots, Some(&txn))
```

Comment:
This will fail for historic epochs as well (see comment in reward_registry.
However, I don't think we use this for any old blocks.

Reply:
No, this returns None if `RewardRegistry::get_slot_at` returns `None`

Reply:
But only after your changes ;)